clang-tools-extra/clang-tidy/readability/MathMissingParenthesesCheck.cpp

```diff
@@ -0,0 +1,167 @@
+//===--- MathMissingParenthesesCheck.cpp - clang-tidy ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "MathMissingParenthesesCheck.h"
+#include "../utils/ASTUtils.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/Preprocessor.h"
+#include <set>
+#include <stack>
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+void MathMissingParenthesesCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(binaryOperator(unless(hasParent(binaryOperator())),
+                                    hasDescendant(binaryOperator()))
```

Comment:
Consider generating multiple warnings instead of single one, for example.
For things like: 2*2 - 2/2
we will have
```
   binaryOperator -
          binaryOperator *
          binaryOperator /
```
in such case check could be for binaryOperator(hasParent(binaryOperator().bind("parent"))).bind("current")

if (parent got lower priority than current)
   raise warning and add () to current